File: mmupnp/src/main/java/net/mm2d/upnp/TaskExecutor.java
/*
 * Copyright (c) 2019 大前良介 (OHMAE Ryosuke)
 *
 * This software is released under the MIT License.
 * http://opensource.org/licenses/MIT
 */

package net.mm2d.upnp;

import javax.annotation.Nonnull;

/**
 * タスクを指定スレッドで実行するExecutorのインターフェース
 *
 * @author <a href="mailto:ryo@mm2d.net">大前良介 (OHMAE Ryosuke)</a>
 */
public interface TaskExecutor {
    /**
     * タスクを特定のスレッドで実行する。
     *
     * @param task 実行するタスク
     * @return 実行をできた場合、もしくは実行待ちキューに積むことができた場合true
     */
    boolean execute(@Nonnull Runnable task);

    /**
     * ControlPointの終了処理中にコールされる。
     *
     * スレッドの停止などが必要な場合はここで実装する。
     */
    void terminate();
}

File: mmupnp/src/main/java/net/mm2d/upnp/TaskExecutor.kt
/*
 * Copyright (c) 2019 大前良介 (OHMAE Ryosuke)
 *
 * This software is released under the MIT License.
 * http://opensource.org/licenses/MIT
 */

package net.mm2d.upnp

/**
 * Executor interface that executes tasks in a specified thread.
 *
 * @author [大前良介 (OHMAE Ryosuke)](mailto:ryo@mm2d.net)
 */
interface TaskExecutor {
    /**
     * Execute a task in a specific thread.
     *
     * @param task task to execute
     * @return true if execution could be done or if it could be queued up
     */
    fun execute(task: Runnable): Boolean

    /**
     * Called during ControlPoint exit processing.
     *
     * If need to stop threads etc., implement it here.
     */
    fun terminate() = Unit
}
